src/clean.c

```diff
@@ -0,0 +1,397 @@
+				ret = rmdir(file);
+			} else {
+				ret = unlink(file);
+			}
+			if (ret != 0) {
+				fprintf(stderr, "couldn't remove file %s: %s\n", file, strerror(errno));
+			}
+		}
+		if (ret == 0) {
+			stats.files_removed++;
+		}
+	next:
+		free_string(&file);
+		ret = 0;
+	}
+
+end:
+	closedir(dir);
+	return ret;
+}
+
+static bool is_fullfile(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	return strlen(entry->d_name) == (SWUPD_HASH_LEN - 1);
+}
+
+static bool is_pack_indicator(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	static const char prefix[] = "pack-";
+	static const char suffix[] = ".tar";
+	static const size_t prefix_len = sizeof(prefix) - 1;
+	static const size_t suffix_len = sizeof(suffix) - 1;
+
+	const char *name = entry->d_name;
+	size_t len = strlen(name);
+	if (len < (prefix_len + suffix_len)) {
+		return false;
+	}
+	return !memcmp(name, prefix, prefix_len) && !memcmp(name + len - suffix_len, suffix, suffix_len);
+}
+
+static bool is_all_digits(const char *s)
+{
+	while (*s) {
+		if (!isdigit(*s)) {
+			return false;
+		}
+		s++;
+	}
+	return true;
+}
+
+static bool is_manifest(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	static const char prefix[] = "Manifest.";
+	static const size_t prefix_len = sizeof(prefix) - 1;
+
+	const char *name = entry->d_name;
+	if (strlen(name) < prefix_len) {
+		return false;
+	}
+	return !memcmp(entry->d_name, prefix, prefix_len);
+}
+
+static char *read_mom_contents(int version)
+{
+	char *mom_path = NULL;
+	char *contents = NULL;
+
+	string_or_die(&mom_path, "%s/%d/Manifest.MoM", state_dir, version);
+	FILE *f = fopen(mom_path, "r");
+	if (!f) {
+		/* This is a best effort. */
+		goto end;
+	}
+
+	int ret;
+	ret = fseek(f, 0, SEEK_END);
+	if (ret != 0) {
+		goto close_then_end;
+	}
+
+	long size = ftell(f);
+	/* Don't care about failures or empty file. */
+	if (size < 1) {
+		goto end;
+	}
+
+	ret = fseek(f, 0, SEEK_SET);
+	if (ret != 0) {
+		goto close_then_end;
+	}
+
+	char *tmp = malloc(size + 1);
+	if (!tmp) {
+		goto close_then_end;
+	}
+
+	ret = fread(tmp, size, 1, f);
+	if (ret != 1) {
+		free_string(&tmp);
+		goto close_then_end;
+	}
+
+	contents = tmp;
+
+close_then_end:
+	fclose(f);
+
+end:
+	free_string(&mom_path);
+	return contents;
+}
+
+static int clean_staged_manifests()
+{
+	int ret = 0;
+	DIR *dir;
+
+	dir = opendir(state_dir);
+	if (!dir) {
+		ret = errno;
+		goto end;
+	}
+
+	/* NOTE: Currently Manifest files have their timestamp from generation
+	 * preserved. */
+
+	/* When --all is not used, keep the Manifests used by the current OS version. This
+	 * ensures that a regular 'clean' won't make 'search' redownload files. */
+	char *mom_contents = NULL;
+	if (!options.all) {
+		int current_version = get_current_version(path_prefix);
+		if (current_version < 0) {
+			fprintf(stderr, "Unable to determine current OS version\n");
+		} else {
+			mom_contents = read_mom_contents(current_version);
+		}
+	}
+
+	while (true) {
```

Comment:
while (ret == 0) perhaps?

Reply:
Kept as is. It did not look very good in practice since the test doesn't cover all the breaking conditions (we have a case where the stream is over, and ret == 0).